Reorganize contracts

istankovic opened the pull request:
Part of the Arbitrum integration work, this PR prepares the way for Arbitrum-specific contracts by introducing a new contract organization. The new directory structure:
```
contracts/
├── BeamerUtils.sol
├── chains
│   ├── optimism
│   │   ├── ICrossDomainMessenger.sol
│   │   ├── Lib_PredeployAddresses.sol
│   │   ├── OptimismMessengers.sol
│   │   └── OptimismProofSubmitter.sol
│   └── test
│       ├── TestMessengers.sol
│       └── TestProofSubmitter.sol
├── CrossDomainRestrictedCalls.sol
├── FillManager.sol
├── MintableToken.sol
├── RequestManager.sol
├── ResolutionRegistry.sol
├── Resolver.sol
├── RestrictedCalls.sol
└── StandardToken.sol
```

![image](https://user-images.githubusercontent.com/1312091/170292287-d68459c8-5418-41e7-9e0e-279a2928b56f.png)


Red lines indicate restricted calls, red dashed lines indicate restricted calls that cross the L1/L2 boundary.
Black lines represent ordinary, unrestricted calls.
With the new scheme we would have 3 rollup-specific contracts: `L1Messenger` (deployed on L1), `L2Messenger` (deployed on L2) and `ProofSubmitter` (deployed on L2).
Everything else would be rollup-independent.

Closes #662.

fredo commented on contracts/contracts/chains/optimism/OptimismMessengers.sol:
```diff
@@ -0,0 +1,38 @@
+    }
+}
+
+contract OptimismL2Messenger is OptimismMessengerBase {
```
Why do we need the distinction between L1 and L2 messenger? I thought you wrote it in a way that it is equal. Is it just to have the predeployed address?

istankovic replied:
For the Optimism case, both are almost identicall, but as we go on adding support for other rollups I suspect we might run into more differences. So I thought it was good to leave some space for that possibility and to clearly indicate which messengers belong to which chain.

fredo commented on scripts/deployment/rinkeby_testnets.json:
```diff
@@ -9,14 +9,16 @@
             "chain_id": 588,
-            "messenger_contract_address": "0xBa6f1aAF3F909F208b224F3985eCda745FA9101A",
+            "proof_submitter": "OptimismProofSubmitter",
+            "l2_to_l1_messenger": "0xBa6f1aAF3F909F208b224F3985eCda745FA9101A",
```
here is some inconsistent naming. Anywhere else the term `l2_messenger` or `l1_messenger` is used. I think consistency would be nicer.

istankovic replied:
Good catch! That's a leftover from a previous iteration of the patches and is clearly wrong.

istankovic replied:
Fixed.

palango commented on .gitmodules:
```diff
@@ -1,5 +1,5 @@
 [submodule "docker/optimism"]
-	path = docker/optimism
+	path = docker/optimism/optimism
```
why change this?

istankovic replied:
> why change this?

This makes it consistent with the arbitrum stuff and also leaves the room for optimism-specific scripts in `docker/optimism`, like the one we already have, `optimism.sh`.

I also thought about moving all that stuff into e.g. `${ROOT}/integrations/` or `${ROOT}/rollups/` or something similar, because it does not have anything to do with our docker setup, as the toplevel dir name may suggest.

palango commented on .solhintignore:
```diff
@@ -0,0 +1,2 @@
+contracts/contracts/chains/optimism/ICrossDomainMessenger.sol
+contracts/contracts/chains/optimism/Lib_PredeployAddresses.sol
```
Why do we ignore those?

palango commented on contracts/contracts/chains/optimism/Lib_PredeployAddresses.sol:
```diff
@@ -0,0 +1,21 @@
+/**
+ * @title Lib_PredeployAddresses
+ */
+library Lib_PredeployAddresses {
```
Do we expect this to change (often) in the future?

istankovic replied:
I guess the values here will not change anytime soon, but I nevertheless wanted to have meaningful, searchable names instead of simply a long constant. I think this way makes it more clear where the value came from and would make it easier to search for in the future.

fredo commented on docs/source/contracts.dot:
```diff
@@ -0,0 +1,50 @@
+digraph "Beamer contracts" {
```
Can you maybe also mark chain-independent contracts and chain dependent contracts with different colors?

In general it would be nice to have clearly explained somewhere (in one or two sentences) that there are chain dependent contracts to work with their specific messengers (or more abstract contract interfaces) and chain independent ones which are purely dependent to our protocol. Basically a short description why we need to distinguish.

istankovic replied:
I totally agree and I planned to write a bit more, but my time was eaten away by... more pressing concerns. :sweat_smile: 
Now that the Optimism E2E test is working, I can write some more.

istankovic replied:
Done. Chain-dependent contracts are now colored differently. Also added tables listing all the contracts and the chains they are deployed to. Finally, some wording has been improved and a few sentences added to make things more clear.

fredo commented on docker/optimism/optimism.sh:
```diff
@@ -1,61 +1,70 @@
 #!/usr/bin/env bash
-ROOT=$(dirname $(dirname "$(realpath "$0")"))
-DOCKER_COMPOSE_FILES="-f ${ROOT}/beamer/docker/optimism/ops/docker-compose.yml -f ${ROOT}/beamer/docker/optimism/ops/docker-compose-nobuild.yml"
+. "$(realpath $(dirname $0))/../scripts/common.sh"
```
very nice composition of tasks :)

istankovic replied:
Eh... bash... :-)

palango commented on contracts/contracts/chains/test/TestMessengers.sol:
```diff
@@ -0,0 +1,43 @@
+    }
+}
+
+contract TestL1Messenger is TestMessengerBase {}
```
Can you suppress those warning below_

istankovic replied:
Fixed.

palango commented on docker/optimism/optimism.sh:
```diff
@@ -0,0 +1,101 @@
+up() {
+    echo "Starting the end-to-end environment"
+    docker_compose_file="-f ${OPTIMISM}/ops/docker-compose.yml"
+    docker compose ${docker_compose_file} up --scale relayer=1 -d
```
If this doesn't work, why have `--scale relayer=1` at all?

istankovic replied:
Good point, removed.

palango commented on scripts/deployment/main.py:
```diff
@@ -67,7 +68,18 @@ def contract_to_json(contracts: dict[str, DeployedContract]) -> dict:
 
-def deploy_contract(web3: Web3, name: str, *args: Any) -> DeployedContract:
+def deploy_contract(web3: Web3, constructor_spec: Union[str, Sequence]) -> DeployedContract:
+    # constructor_spec is either
```
If this is so, then the type annotation is wrong.

istankovic replied:
> If this is so, then the type annotation can be improved

I'm not sure what you mean, could you please expand?
If you meant something like

```python
constructor_spec: Union[str, Sequence[str]]
```

that would unfortunately not work because there are types other than `str` that may appear there.
Maybe something like `Union[str, Sequence[Union[str, int, ChecksumAddress]]]` might work, but I haven't tried...

palango commented on scripts/deployment/main.py:
```diff
@@ -196,56 +209,39 @@ def main(
 ) -> None:
-    commit_id = get_commit_id()
     account = account_from_keyfile(keystore_file, password)
+    print("Deployer:", account.address)
```
This should probably go away

istankovic replied:
Actually, I added this intentionally since it proved useful when I was inspecting the output of the deployment script and correlating that with the E2E script. It may come in handy so I'd prefer to leave it in, but if you feel strongly about it, I can remove it.

istankovic commented:
@palango could you please take a look at the 3 newly added commits? I've added these because I noticed the deployment script was flaky on CI -- we need to wait for the tx receipt. For that I reused the existing code in the agent, just moved it to `beamer.util` and made it public.
